File: j2kt/jre/java/common/java/lang/Short.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package java.lang;

import static javaemul.internal.KtNativeUtils.ktNative;

import javaemul.internal.annotations.KtNative;
import org.jspecify.nullness.NullMarked;
import org.jspecify.nullness.Nullable;

@KtNative(name = "kotlin.Short", companionName = "java.lang.Short")
@NullMarked
public final class Short extends Number implements Comparable<Short> {

  public static final short MAX_VALUE = 32767;

  public static final short MIN_VALUE = -32768;

  public static final int SIZE = 16;

  public static final int BYTES = SIZE / Byte.SIZE;

  public static final Class<Short> TYPE = ktNative();

  public Short(String string) throws NumberFormatException {}

  public Short(short value) {}

  @Override
  public native byte byteValue();

  @Override
  public native int compareTo(Short object);

  public static native int compare(short lhs, short rhs);

  public static native Short decode(String string) throws NumberFormatException;

  @Override
  public native double doubleValue();

  @Override
  public native boolean equals(@Nullable Object object);

  @Override
  public native float floatValue();

  @Override
  public native int hashCode();

  @Override
  public native int intValue();

  @Override
  public native long longValue();

  public static native short parseShort(String string) throws NumberFormatException;

  public static native short parseShort(String string, int radix) throws NumberFormatException;

  @Override
  public native short shortValue();

  @Override
  public native String toString();

  public static native String toString(short value);

  public static native Short valueOf(String string) throws NumberFormatException;

  public static native Short valueOf(String string, int radix) throws NumberFormatException;

  public static native short reverseBytes(short s);

  public static native Short valueOf(short s);

  public static native int hashCode(short s);
}

File: j2kt/jre/java/native/java/lang/Short.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.lang

import javaemul.internal.decodeNumberString
import kotlin.jvm.javaPrimitiveType

/** java.lang.Short static method emulations */
object Short {
  operator fun invoke(s: kotlin.Short): kotlin.Short = s

  operator fun invoke(s: kotlin.String): kotlin.Short = s.toShort()

  const val MAX_VALUE = kotlin.Short.MAX_VALUE

  const val MIN_VALUE = kotlin.Short.MIN_VALUE

  const val SIZE = kotlin.Short.SIZE_BITS

  const val BYTES = kotlin.Short.SIZE_BYTES

  val TYPE: Class<kotlin.Short> = kotlin.Short::class.javaPrimitiveType!!

  fun valueOf(s: kotlin.Short): kotlin.Short = s

  fun valueOf(str: kotlin.String): kotlin.Short = str.toShort()

  fun valueOf(s: kotlin.String, radix: Int): kotlin.Short = s.toShort(radix)

  fun compare(s1: kotlin.Short, s2: kotlin.Short): Int = s1.compareTo(s2)

  fun decode(s: kotlin.String): kotlin.Short {
    val (radix, payload) = decodeNumberString(s)
    return payload.toShort(radix)
  }

  fun toString(s: kotlin.Short): kotlin.String = s.toString()

  fun parseShort(str: kotlin.String): kotlin.Short = str.toShort()

  fun parseShort(s: kotlin.String, radix: Int): kotlin.Short = s.toShort(radix)

  fun hashCode(s: kotlin.Short): Int = s.hashCode()

  fun reverseBytes(s: kotlin.Short): kotlin.Short =
    (((s.toInt() shr 8) and 0xff) or ((s.toInt() and 0xff) shl 8)).toShort()
}
